flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java

```diff
@@ -186,10 +188,18 @@ private void maybeHandleTaskFailure(final TaskExecutionState taskExecutionState,
 
 	private void handleTaskFailure(final ExecutionVertexID executionVertexId, @Nullable final Throwable error) {
 		setGlobalFailureCause(error);
+		notifyCoordinatorsAboutTaskFailure(executionVertexId, error);
```

Comment:
It looks that the failover may initially restart tasks only and fall back to a global failover if the task-only failover encountered exception. This may cause the `OperatorCoordinator.subtaskFailed()` to be invoked twice consecutively on the same subtask. Currently `SourceCoordinator` throws an exception if `subtaskFailed()` is invoked on a subtask that has not been registered. We may need to remove that check.

Reply:
I think we may need to remove that check anyways. It is quite possible that tasks fail immediately after they were deployed, before they could register.

The `OperatorCoordinator` interface also has no notion of "registering", so it cannot decide which notifications to forward and which not. I'd rather err on the "forward too many" side, because missing a notification easily leads to a stall or data loss (split not added back, result is missing).

Reply:
Sounds good to me. Maybe we can just remove it in this PR. The code is here:
https://github.com/apache/flink/blob/6f64213dbdfd0b1e75f503007dd026162881c9c7/flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java#L252